<html>
<head>
<title>A Beginner’s Guide to Debugging Code Using Python</title>
<link href="../Styles/Style.css" type="text/css" rel="stylesheet"/>
</head>
<body>
<h1 class="translated">使用Python调试代码的初学者指南</h1>
<blockquote>原文：<a href="https://levelup.gitconnected.com/a-beginners-guide-to-debugging-code-using-python-911ea76380fc?source=collection_archive---------19-----------------------#2022-05-02">https://levelup.gitconnected.com/a-beginners-guide-to-debugging-code-using-python-911ea76380fc?source=collection_archive---------19-----------------------#2022-05-02</a></blockquote><div><div class="fc ih ii ij ik il"/><div class="im in io ip iq"><div class=""/><figure class="gl gn jr js jt ju gh gi paragraph-image"><div role="button" tabindex="0" class="jv jw di jx bf jy"><div class="gh gi jq"><img src="../Images/f153fddb08a96bed3609f8d2fb62a781.png" data-original-src="https://miro.medium.com/v2/resize:fit:1400/format:webp/1*2rZ5AVuaAD3btjHmdTTk8w.jpeg"/></div></div><figcaption class="kb kc gj gh gi kd ke bd b be z dk translated">米格尔·Á拍摄的照片。来自<a class="ae kf" href="https://www.pexels.com/photo/close-up-shot-of-keyboard-buttons-2882552/?utm_content=attributionCopyText&amp;utm_medium=referral&amp;utm_source=pexels" rel="noopener ugc nofollow" target="_blank">佩克斯</a>的帕德里纳</figcaption></figure><p id="66b8" class="pw-post-body-paragraph kg kh it ki b kj kk kl km kn ko kp kq kr ks kt ku kv kw kx ky kz la lb lc ld im bi translated">你有没有注意到你自己试图从头开始学习编码，但是相反，你花了几个小时盯着一个空白页或者无休止地浏览你最终确定的工作流程？原因可能是当您看到屏幕底部的红色粗体错误文本时，内心深处会有一些恐惧，有时会感到非常沮丧。尽管如此，作为一个经历了四年计算机科学课程并在技术行业工作了三年多的人，我想让大家知道，从构建阶段的开始到结束，看到编码错误是正常的。我将通过简单的例子来介绍一些调试策略，希望可以帮助你或任何试图进入编码领域的新手。</p><h1 id="b1a2" class="le lf it bd lg lh li lj lk ll lm ln lo lp lq lr ls lt lu lv lw lx ly lz ma mb bi translated">注意Python中区分大小写的语法:</h1><p id="5074" class="pw-post-body-paragraph kg kh it ki b kj mc kl km kn md kp kq kr me kt ku kv mf kx ky kz mg lb lc ld im bi translated">语法错误是我们的软肋，尤其是如果你声明了很多变量，而这些变量最终可能会因为出现的错误而遗漏。语法错误通常是打字错误、缺少括号和逗号。随着代码变得越来越长，你必须非常小心，并且你需要保持一个有组织的结构。我喜欢在记事本上写下变量名的列表，然后写一个简单的注释，说明这个变量在代码中做了什么。您可以通过对代码进行注释来做到这一点，这使得下一个接手代码的开发人员更加容易。</p><h2 id="2628" class="mh lf it bd lg mi mj dn lk mk ml dp lo kr mm mn ls kv mo mp lw kz mq mr ma ms bi translated">示例显示了Python中的一个简单错误</h2><pre class="mt mu mv mw gt mx my mz na aw nb bi"><span id="770b" class="mh lf it my b gy nc nd l ne nf">print "Hello World!"</span></pre><h2 id="a7e2" class="mh lf it bd lg mi mj dn lk mk ml dp lo kr mm mn ls kv mo mp lw kz mq mr ma ms bi translated">输出</h2><pre class="mt mu mv mw gt mx my mz na aw nb bi"><span id="9ed4" class="mh lf it my b gy nc nd l ne nf">line 1<br/>    print "Hello World!"<br/>          ^<br/>SyntaxError: Missing parentheses in call to 'print'. Did you mean print("Hello World!")?</span></pre><h1 id="37d5" class="le lf it bd lg lh li lj lk ll lm ln lo lp lq lr ls lt lu lv lw lx ly lz ma mb bi translated"><strong class="ak">饲养蟒蛇异常:</strong></h1><p id="1a74" class="pw-post-body-paragraph kg kh it ki b kj mc kl km kn md kp kq kr me kt ku kv mf kx ky kz mg lb lc ld im bi translated">什么是例外？异常是在程序执行期间发生的干扰程序流程的事件。本质上，当代码运行时的开始阶段出现异常时，它将停止其余代码的执行。那么我们如何在Python中引发异常呢？在代码中，可以通过在条件语句中使用关键字<em class="ng">“raise”</em>来使用<a class="ae kf" href="https://www.w3schools.com/python/ref_keyword_raise.asp" rel="noopener ugc nofollow" target="_blank">引发异常</a>。当您<em class="ng"/>引发一个异常时，您可以定义在条件语句不满足时应该出现的硬编码文本。如果你很好奇，可以点击这里的链接<a class="ae kf" href="https://www.tutorialsteacher.com/python/error-types-in-python" rel="noopener ugc nofollow" target="_blank">查看一些你可能在代码中看到的常见异常错误。</a></p><h2 id="077a" class="mh lf it bd lg mi mj dn lk mk ml dp lo kr mm mn ls kv mo mp lw kz mq mr ma ms bi translated">示例显示了用于错误处理的<em class="nh">“raise”</em>语法</h2><pre class="mt mu mv mw gt mx my mz na aw nb bi"><span id="3cc4" class="mh lf it my b gy nc nd l ne nf">z = "This is not an integer!"<br/><br/>if not type(z) is int:<br/>  raise TypeError("Only integers are allowed")</span></pre><h2 id="f7db" class="mh lf it bd lg mi mj dn lk mk ml dp lo kr mm mn ls kv mo mp lw kz mq mr ma ms bi translated">输出</h2><pre class="mt mu mv mw gt mx my mz na aw nb bi"><span id="ca5e" class="mh lf it my b gy nc nd l ne nf">line 4, in &lt;module&gt;<br/>    raise TypeError("Only integers are allowed")<br/>TypeError: Only integers are allowed</span></pre><h1 id="b39f" class="le lf it bd lg lh li lj lk ll lm ln lo lp lq lr ls lt lu lv lw lx ly lz ma mb bi translated">捕捉和处理Python异常:</h1><p id="4f90" class="pw-post-body-paragraph kg kh it ki b kj mc kl km kn md kp kq kr me kt ku kv mf kx ky kz mg lb lc ld im bi translated">Python中处理异常有什么用？如果在代码运行时没有正确处理异常，程序将会崩溃。处理异常是很重要的，尤其是当你的程序有多个相互调用的函数时。捕捉异常的最佳实践是使用<a class="ae kf" href="https://www.w3schools.com/python/python_try_except.asp" rel="noopener ugc nofollow" target="_blank"><em class="ng">【try】</em>和<em class="ng">【except】</em>子句</a>。<em class="ng">“try”</em>子句允许您测试代码中的错误，然后在其后的<em class="ng">“except”</em>子句允许您处理错误。这是在运行编译器时处理意外错误的好方法。</p><h2 id="1730" class="mh lf it bd lg mi mj dn lk mk ml dp lo kr mm mn ls kv mo mp lw kz mq mr ma ms bi translated">示例显示了异常的处理</h2><pre class="mt mu mv mw gt mx my mz na aw nb bi"><span id="c26e" class="mh lf it my b gy nc nd l ne nf">try:<br/> print(x)<br/>except:<br/> print(“x is not declared so an exception occurred.”)</span></pre><h2 id="a989" class="mh lf it bd lg mi mj dn lk mk ml dp lo kr mm mn ls kv mo mp lw kz mq mr ma ms bi translated">输出</h2><pre class="mt mu mv mw gt mx my mz na aw nb bi"><span id="283e" class="mh lf it my b gy nc nd l ne nf">x is not declared so an exception occurred.</span></pre><h2 id="7f10" class="mh lf it bd lg mi mj dn lk mk ml dp lo kr mm mn ls kv mo mp lw kz mq mr ma ms bi translated">示例显示了内置异常(ValueError)</h2><pre class="mt mu mv mw gt mx my mz na aw nb bi"><span id="732a" class="mh lf it my b gy nc nd l ne nf">while True:<br/>    try:<br/>        z = int(input("Please enter a number: "))<br/>    except ValueError:<br/>        print("That was not a valid number.  Try again!")</span></pre><h2 id="c797" class="mh lf it bd lg mi mj dn lk mk ml dp lo kr mm mn ls kv mo mp lw kz mq mr ma ms bi translated">输出</h2><pre class="mt mu mv mw gt mx my mz na aw nb bi"><span id="3500" class="mh lf it my b gy nc nd l ne nf">Please enter a number: 1<br/>Please enter a number: 12<br/>Please enter a number: Hello!<br/>That was not a valid number.  Try again!<br/>Please enter a number:</span></pre><h1 id="d1c8" class="le lf it bd lg lh li lj lk ll lm ln lo lp lq lr ls lt lu lv lw lx ly lz ma mb bi translated">在Python中断言:</h1><p id="de63" class="pw-post-body-paragraph kg kh it ki b kj mc kl km kn md kp kq kr me kt ku kv mf kx ky kz mg lb lc ld im bi translated">Python中的assert语句将检查条件是否为真。<a class="ae kf" href="https://www.w3schools.com/python/ref_keyword_assert.asp" rel="noopener ugc nofollow" target="_blank"><em class="ng">【assert】</em>语法</a>可以接受两个参数:条件和消息。假设条件为真，该语句用于继续执行。但是如果assert语句为false，Python将引发<em class="ng">“assertion error”</em>异常，并给出指定的错误消息。</p><h2 id="8604" class="mh lf it bd lg mi mj dn lk mk ml dp lo kr mm mn ls kv mo mp lw kz mq mr ma ms bi translated">示例显示了如何使用assert来检查条件是否为真</h2><pre class="mt mu mv mw gt mx my mz na aw nb bi"><span id="62b6" class="mh lf it my b gy nc nd l ne nf">z = 1<br/>assert z &gt; 0<br/>print('z is a positive number.')</span></pre><h2 id="6b26" class="mh lf it bd lg mi mj dn lk mk ml dp lo kr mm mn ls kv mo mp lw kz mq mr ma ms bi translated">输出</h2><pre class="mt mu mv mw gt mx my mz na aw nb bi"><span id="4155" class="mh lf it my b gy nc nd l ne nf">z is a positive number.</span></pre><h2 id="cb49" class="mh lf it bd lg mi mj dn lk mk ml dp lo kr mm mn ls kv mo mp lw kz mq mr ma ms bi translated">示例显示变量不满足条件，因此它将返回AssertionError</h2><pre class="mt mu mv mw gt mx my mz na aw nb bi"><span id="26b9" class="mh lf it my b gy nc nd l ne nf">z = -1<br/>assert z &gt; 0<br/>print('z is a positive number.')</span></pre><h2 id="2668" class="mh lf it bd lg mi mj dn lk mk ml dp lo kr mm mn ls kv mo mp lw kz mq mr ma ms bi translated">输出</h2><pre class="mt mu mv mw gt mx my mz na aw nb bi"><span id="fa66" class="mh lf it my b gy nc nd l ne nf">line 2, in &lt;module&gt;<br/>    assert z &gt; 0<br/>AssertionError</span></pre><h1 id="a4b4" class="le lf it bd lg lh li lj lk ll lm ln lo lp lq lr ls lt lu lv lw lx ly lz ma mb bi translated">在Python中登录:</h1><p id="c16c" class="pw-post-body-paragraph kg kh it ki b kj mc kl km kn md kp kq kr me kt ku kv mf kx ky kz mg lb lc ld im bi translated">对于任何经历测试阶段的开发人员来说，日志记录都是最有价值的工具之一，甚至可以帮助您发现在最初设计时从未想过的新测试场景。幸运的是，Python有一个内置的日志模块，可以通过键入<em class="ng">“导入日志”轻松地导入到您的代码中</em>这里有一个很好的<a class="ae kf" href="https://docs.python.org/3/library/logging.html" rel="noopener ugc nofollow" target="_blank">链接</a>到关于登录Python可以做什么的文档。通过利用<em class="ng">“基本配置”</em>配置，您可以定制日志记录以满足您的需求，在这里您可以决定文件名、文件格式、日期格式和严重性级别(严重、错误、警告、信息、调试和不设置)。</p><h2 id="04af" class="mh lf it bd lg mi mj dn lk mk ml dp lo kr mm mn ls kv mo mp lw kz mq mr ma ms bi translated">如何在Python中设置简单记录器的示例</h2><pre class="mt mu mv mw gt mx my mz na aw nb bi"><span id="29e4" class="mh lf it my b gy nc nd l ne nf">import logging<br/><br/>logger = logging.getLogger()<br/>logger.setLevel(logging.DEBUG)<br/>logging.basicConfig(filename="test.log", format='%(asctime)s %(filename)s: %(message)s', datefmt='%H:%M:%S')<br/>logger.info("testing information message1!")</span></pre><h2 id="7be8" class="mh lf it bd lg mi mj dn lk mk ml dp lo kr mm mn ls kv mo mp lw kz mq mr ma ms bi translated">输出</h2><p id="811c" class="pw-post-body-paragraph kg kh it ki b kj mc kl km kn md kp kq kr me kt ku kv mf kx ky kz mg lb lc ld im bi translated">生成名为test.log的文本文件</p><h1 id="f333" class="le lf it bd lg lh li lj lk ll lm ln lo lp lq lr ls lt lu lv lw lx ly lz ma mb bi translated">结论:</h1><p id="a691" class="pw-post-body-paragraph kg kh it ki b kj mc kl km kn md kp kq kr me kt ku kv mf kx ky kz mg lb lc ld im bi translated">我希望这篇文章对编程新手和试图理解调试代码中出现的问题的最佳方式的人有所帮助。我希望这可以作为你的垫脚石，并变得更加舒适地处理可能意想不到的错误。</p></div></div>    
</body>
</html>